packages/roosterjs-content-model-plugins/lib/markdown/utils/setFormat.ts

```diff
@@ -0,0 +1,90 @@
+import { createText, getSelectedSegmentsAndParagraphs } from 'roosterjs-content-model-dom';
+import type {
+    ContentModelSegmentFormat,
+    ContentModelText,
+    IEditor,
+} from 'roosterjs-content-model-types';
+
+/**
+ * @internal
+ */
+export function setFormat(editor: IEditor, character: string, format: ContentModelSegmentFormat) {
+    editor.formatContentModel((model, context) => {
+        const selectedSegmentsAndParagraphs = getSelectedSegmentsAndParagraphs(
+            model,
+            false /*includeFormatHolder*/
+        );
+
+        if (selectedSegmentsAndParagraphs.length > 0 && selectedSegmentsAndParagraphs[0][1]) {
+            const marker = selectedSegmentsAndParagraphs[0][0];
+            context.newPendingFormat = {
+                ...marker.format,
+                strikethrough: !!marker.format.strikethrough,
+                italic: !!marker.format.italic,
+                fontWeight: marker.format?.fontWeight ? 'bold' : undefined,
+            };
+
+            const paragraph = selectedSegmentsAndParagraphs[0][1];
+            if (marker.segmentType == 'SelectionMarker') {
+                const markerIndex = paragraph.segments.indexOf(marker);
+                let previousCharacterSegment: ContentModelText | undefined = undefined;
+                let index = markerIndex;
+                if (markerIndex > 0 && paragraph.segments[markerIndex - 1]) {
+                    const segmentBeforeMarker = paragraph.segments[markerIndex - 1];
+                    if (
+                        segmentBeforeMarker.segmentType == 'Text' &&
+                        segmentBeforeMarker.text[segmentBeforeMarker.text.length - 1] == character
+                    ) {
+                        for (let i = markerIndex - 1; i >= 0; i--) {
+                            const previousSegment = paragraph.segments[i];
+                            if (
+                                previousSegment &&
+                                previousSegment.segmentType == 'Text' &&
+                                previousSegment.text.indexOf(character) > -1 &&
+                                !(
+                                    i == markerIndex - 1 &&
+                                    previousSegment.text.indexOf(character) ==
+                                        previousSegment.text.length - 1
+                                )
+                            ) {
+                                previousCharacterSegment = previousSegment;
+                                index = i;
+                                break;
+                            }
+                        }
+                        if (previousCharacterSegment && index < markerIndex) {
+                            const splittedSegments = previousCharacterSegment.text.split(character);
```

Comment:
what if the result has more than 2 segments?

For example "a*b*c*d", it will return an array of 4.

Reply:
It will search for the segment that has * and it will iterate through all the segments until the the second *. The for in line 38 search for the segments and the for in line 71 will apply the formatting.

Reply:
I remove the code to handle multiple segments, since the legacy markdown do not handle multiple styles ate the same time.